kernel: backport netfilter-actual-sk (19.07)

aaronjg opened the pull request:
Backport of @zx2c4's  linux kernel commit 46d6c5a to 4.14 kernel.

netfilter: use actual socket sk rather than skb sk when routing harder

aaronjg committed:
kernel: add netfilter-actual-sk patch

Backport of linux kernel commit 46d6c5a to 4.14 kernel.

netfilter: use actual socket sk rather than skb sk when routing harder

Signed-off-by: Aaron Goodman <aaronjg@stanford.edu>

adschm commented:
If that's a backport, you should definitely keep the patch header in the patch file.

Is this already included in the 5.4 kernel we use in master?

zx2c4 commented:
@aaronjg I backported that to 4.19 the other day and posted it to stable@ and @gregkh picked it up. If this patch is correct (I haven't looked yet), can you send it to stable@?

aaronjg commented:
@adschm: Okay, I just added back the original patch header and added an additional signoff for the backport. I built the patch using quilt, so it was not included initially.

It looks like it [has been backported](https://www.spinics.net/lists/stable/msg426855.html) to 5.4, though I don't think it is in the version that we are using in master. I assumed that by the time we get to the 20.x release, we will be on a kernel version that has the patch included. I have tested @zx2c4's original patch on 5.4, and it does work.

@zx2c4: Thanks. I wasn't aware you had already done the 4.19 backport. That would have made this easier.

There was a change to  `nf_ip6_reroute` in [308-v4.16-netfilter-move-reroute-indirection-to-struct-nf_ipv6.patch](https://github.com/openwrt/openwrt/blob/openwrt-19.07/target/linux/generic/backport-4.14/308-v4.16-netfilter-move-reroute-indirection-to-struct-nf_ipv6.patch), so I don't think this will apply cleanly to the unpatched 4.14.

ynezz commented:
Merged, thanks!
